### PR TITLE
Color sub

### PR DESCRIPTION
**Pull request by godlzr**

### CODE REVIEW

**Comment by WestLangley** [comment]
This can yield negative color components. I am not sure that is what we want. What is your use case for this?

@mrdoob Do you want to go as far as adding blending methods to the `Color` class?

    Color.blend( color, operation );

**Comment by godlzr** [comment]
Sometimes, you may want to decompose a color and get the original two color, like material conversion between different shader models. Not sure how to handle the negative case, but how do you handle the value greater than 255?

**Comment by WestLangley** [comment]
I was asking about your use case... Do you mind being more specific?

Valid `Color` components are nominally from 0 to 1.

**Comment by godlzr** [comment]
It's simple, I have two kinds of shading models, A.specular = B.specular + B.metallic  + sth.
when I do inverse calculation for conversion, I need to use sub().
I am not sure if sub method is really useful for others, if not I can close the PR

**Comment by mrdoob** [comment]
The same way we do `min( 1, value )` we can probably do `max( 0, value )`. So this sounds good to me.

File: docs/api/math/Color.html [hunk 1]
```diff
@@ -185,6 +185,11 @@
 		Adds given h, s, and l to this color's existing h, s, and l values.
 		</div>
 
+		<h3>[method:Color sub]( [page:Color color] ) [page:Color this]</h3>
+		<div>
+		Subs rgb values of given color to rgb values of this color
```
Comment (WestLangley) [inline docs/api/math/Color.html]: Subtracts rgb components of given color from rgb components of this color. If a component is negative, it is set to zero.

File: src/math/Color.js [hunk 2]
```diff
@@ -381,6 +381,16 @@ THREE.Color.prototype = {
 
 	},
 
+	sub: function( color ) {
+
+		this.r -= color.r;
+		this.g -= color.g;
```
Comment (WestLangley) [inline src/math/Color.js]: this.r = Math.max( 0, this.r - color.r );

**Comment by bhouston** [comment]
BTW @godlzr contributed this because we needed it for https://Clara.io.  I think the max() clamping approach is a nice solution.

**Comment by bhouston** [comment]
@WestLangley:
> Valid Color components are nominally from 0 to 1.

That is true by convention, but it isn't actually true of physical colors.

Reflectivity of a surface is limited to the range of 0-1, thus colors that represents the reflectivity coefficients of a surface are limited to 0-1.  Convention also states that the range 0-1 can be mapped directly to screen intensity.

But colors representing light can be anything between 0 and infinite.  And inside the glsl code we regularly let colors exceed the 0-1 range.

It would be so cool if we only clamped to the range 0-1 when converting to hex values or other representations that have range limitations - like the glsl code does.  Otherwise people have to clamp the colors themselves if they wanted to restrict themselves to reflectivity coefficients or LDR colors -- like the glsl code does.

It would make the code simpler and more general purpose.

**Comment by WestLangley** [comment]
@bhouston Yes, I know. I was responding to @godlzr's question regarding values "greater than 255".

In passing data to the GPU, we are confounding "color" which is unit-less, and light "intensity", which has units particular to the type of light. This was presumably to save on uniforms.

I intend to work with you on all this soon -- as I free up time. But yes, we represent only LDR colors on the CPU side. Suggestions on how to improve the API would be worth exploring.

**Comment by bhouston** [comment]
In my opinion, Color is ambiguous and not well defined and its meaning is usually context dependent.

Reflectivity, transmissive, absoprtion, emission, irradiance, radiance, power, etc are all well defined physical properties that can be represented using a color class.  Some of these physical properties have different valid ranges.  I'd like to see a single color class that can represent all of them.

**Comment by bhouston** [comment]
I guess we have a single Vector3 class that can represent both normals (which should always have a length of 1), relative delta as well as positions in space.  I am very glad that we do not have a THREE.Normal3, a THREE.Point3 in addition to the THREE.Vector3 -- although I have seen this pattern before, e.g. https://github.com/stephan-rehfeld/scaladelray/tree/master/src/main/scala/scaladelray/math https://github.com/K-3D/k3d/tree/master/k3dsdk

I guess in a perfect world I'd like to see both Color3 and Color4 as well.  I think that it is not advantageous to always use a Color4 for cases where a Color3 is appropriate -- those extra degrees of freedom can create confusion with users and whether an alpha channel is supported.

**Comment by WestLangley** [comment]
@bhouston 

_Reflectance_ is the fraction of incident power that is reflected by a material. It is between 0 and 1, and it is unit-less. We use the `Color` class to represent the diffuse reflectance of the material. This is all consistent.

Some of the other quantities you mentioned do have units and are also unbounded, so a 3-component, rgb-valued, `Vector` class may be best (Just add r,g,b, getters/setters to `Vector3`). I am not sure using `Color` for an unbounded quantity with units is a good idea.

As I have said before, my current view is color is unit-less, and the units come from the light intensity. I know that is not your view, but we need to work this through, and we will. : - )

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**